controllers/istio_extension_reconciler.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/kuadrant/kuadrant-operator/pkg/common"
+	kuadrantistio "github.com/kuadrant/kuadrant-operator/pkg/istio"
+	kuadrantgatewayapi "github.com/kuadrant/kuadrant-operator/pkg/library/gatewayapi"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+	"github.com/kuadrant/kuadrant-operator/pkg/wasm"
+)
+
+// istioExtensionReconciler reconciles Istio WasmPlugin custom resources
+type istioExtensionReconciler struct {
+	client *dynamic.DynamicClient
+}
+
+func (r *istioExtensionReconciler) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.Reconcile,
+		Events: []controller.ResourceEventMatcher{ // matches reconciliation events that change the rate limit definitions or status of rate limit policies
+			{Kind: &kuadrantv1beta1.KuadrantGroupKind},
+			{Kind: &machinery.GatewayClassGroupKind},
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &machinery.HTTPRouteGroupKind},
+			{Kind: &kuadrantv1beta3.RateLimitPolicyGroupKind},
+			{Kind: &kuadrantistio.WasmPluginGroupKind},
+		},
+	}
+}
+
+func (r *istioExtensionReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("istioExtensionReconciler")
+
+	logger.V(1).Info("building istio extension")
+	defer logger.V(1).Info("finished building istio extension")
+
+	// build wasm plugin configs for each gateway
+	wasmConfigs, err := r.buildWasmConfigs(ctx, state)
+	if err != nil {
+		if errors.Is(err, ErrMissingStateEffectiveRateLimitPolicies) {
+			logger.V(1).Info(err.Error())
+		} else {
+			return err
+		}
+	}
+
+	// reconcile for each gateway based on the desired wasm plugin policies calculated before
+	gateways := lo.Map(topology.Targetables().Items(func(o machinery.Object) bool {
+		return o.GroupVersionKind().GroupKind() == machinery.GatewayGroupKind
+	}), func(g machinery.Targetable, _ int) *machinery.Gateway {
+		return g.(*machinery.Gateway)
+	})
+
+	var modifiedGateways []string
+
+	for _, gateway := range gateways {
+		gatewayKey := k8stypes.NamespacedName{Name: gateway.GetName(), Namespace: gateway.GetNamespace()}
+
+		desiredWasmPlugin := buildIstioWasmPluginForGateway(gateway, wasmConfigs[gateway.GetLocator()])
+
+		resource := r.client.Resource(kuadrantistio.WasmPluginsResource).Namespace(desiredWasmPlugin.GetNamespace())
+
+		existingWasmPluginObj, found := lo.Find(topology.Objects().Children(gateway), func(child machinery.Object) bool {
+			return child.GroupVersionKind().GroupKind() == kuadrantistio.WasmPluginGroupKind && child.GetName() == desiredWasmPlugin.GetName() && child.GetNamespace() == desiredWasmPlugin.GetNamespace() && labels.Set(child.(*controller.RuntimeObject).GetLabels()).AsSelector().Matches(labels.Set(desiredWasmPlugin.GetLabels()))
+		})
+
+		// create
+		if !found {
+			if utils.IsObjectTaggedToDelete(desiredWasmPlugin) {
+				continue
+			}
```

Comment:
I am wondering if we restructure the code to have the delete steps before the create can this logic block be removed?